File: src/main/java/cn/solarmoon/idyllic_food_diary/element/matter/cookware/cup/CupBlockEntity.java
package cn.solarmoon.idyllic_food_diary.element.matter.cookware.cup;

import cn.solarmoon.idyllic_food_diary.feature.tea_brewing.IBrewingRecipe;
import cn.solarmoon.idyllic_food_diary.registry.common.IMBlockEntities;
import cn.solarmoon.solarmoon_core.api.tile.SyncedBlockEntity;
import cn.solarmoon.solarmoon_core.api.tile.inventory.TileInventory;
import cn.solarmoon.solarmoon_core.api.tile.fluid.TileTank;
import net.minecraft.core.BlockPos;
import net.minecraft.world.level.block.state.BlockState;
import net.minecraftforge.fluids.capability.templates.FluidTank;
import net.minecraftforge.items.ItemStackHandler;

public class CupBlockEntity extends SyncedBlockEntity implements IBrewingRecipe {

    private final TileInventory inventory;
    private final TileTank fluidTank;


    private int time;
    private int recipeTime;

    public CupBlockEntity(BlockPos pos, BlockState state) {
        super(IMBlockEntities.LITTLE_CUP.get(), pos, state);
        inventory = new TileInventory(1, 1, this);
        fluidTank = new TileTank(250, this);
    }

    @Override
    public void setBrewTime(int time) {
        this.time = time;
    }

    @Override
    public void setBrewRecipeTime(int time) {
        this.recipeTime = time;
    }

    @Override
    public int getBrewTime() {
        return time;
    }

    @Override
    public int getBrewRecipeTime() {
        return recipeTime;
    }

    @Override
    public ItemStackHandler getInventory() {
        return inventory;
    }

    @Override
    public FluidTank getTank() {
        return fluidTank;
    }

}

File: src/main/kotlin/cn/solarmoon/idyllic_food_diary/element/matter/cookware/cup/CupBlockEntity.kt
package cn.solarmoon.idyllic_food_diary.element.matter.cookware.cup

import cn.solarmoon.idyllic_food_diary.element.recipe.BrewingRecipe
import cn.solarmoon.idyllic_food_diary.registry.common.IFDRecipes
import cn.solarmoon.spark_core.api.attachment.animation.AnimHelper
import cn.solarmoon.spark_core.api.blockentity.SyncedBlockEntity
import cn.solarmoon.spark_core.api.cap.fluid.FluidHandlerHelper
import cn.solarmoon.spark_core.api.cap.fluid.TileTank
import cn.solarmoon.spark_core.api.cap.item.ItemStackHandlerHelper
import cn.solarmoon.spark_core.api.cap.item.TileInventory
import cn.solarmoon.spark_core.api.recipe.processor.RecipeProcessorHelper
import cn.solarmoon.spark_core.registry.common.SparkDataComponents
import net.minecraft.core.BlockPos
import net.minecraft.core.Direction
import net.minecraft.core.HolderLookup
import net.minecraft.core.component.DataComponentMap
import net.minecraft.nbt.CompoundTag
import net.minecraft.world.item.ItemStack
import net.minecraft.world.level.block.entity.BlockEntityType
import net.minecraft.world.level.block.state.BlockState
import net.neoforged.neoforge.fluids.SimpleFluidContent

abstract class CupBlockEntity(type: BlockEntityType<*>, size: Int, capacity: Int, pos: BlockPos, state: BlockState): SyncedBlockEntity(type, pos, state) {

    val inventory = object : TileInventory(this, size, 1) {
        override fun isItemValid(slot: Int, stack: ItemStack): Boolean {
            val recipes = level?.recipeManager?.getAllRecipesFor(IFDRecipes.BREWING.type.get()) ?: return false
            return recipes.any { it.value.ingredients.any { it.ingredient.test(stack) } }
        }
    }
    val tank = CupTank(this, capacity)
    val brew = BrewingRecipe.Processor(this, inventory, tank)

    init {
        AnimHelper.Fluid.createFluidAnim(this, Direction.UP)
        RecipeProcessorHelper.createMap(this, brew)
    }

    override fun saveAdditional(tag: CompoundTag, registries: HolderLookup.Provider) {
        super.saveAdditional(tag, registries)
        tag.put(FluidHandlerHelper.FLUID, tank.writeToNBT(registries, CompoundTag()))
        tag.put(ItemStackHandlerHelper.ITEM, inventory.serializeNBT(registries))
    }

    override fun loadAdditional(tag: CompoundTag, registries: HolderLookup.Provider) {
        super.loadAdditional(tag, registries)
        inventory.deserializeNBT(registries, tag.getCompound(ItemStackHandlerHelper.ITEM))
        tank.readFromNBT(registries, tag.getCompound(FluidHandlerHelper.FLUID))
    }

}